doc/asic-fw-update/asic-fw-update-config.md

```diff
@@ -0,0 +1,285 @@
+- Feature should be configurable in SONiC-defined way trough the database.
+- Config should last after image upgrade.
+- Firmware images should last after image upgrade.
+- Have an ability to control auto-update.
+- Have an ability to control firmware source.
+
+### Architecture Design
+This section covers the changes that are required in the SONiC architecture.
+In general, it is expected that the current architecture is not changed.
+This section should explain how the new feature/enhancement (module/sub-module)
+fits in the existing architecture.
+
+### High-Level Design
+#### Original ASIC FW update flow
+The original flow of ASIC firmware update is very siple. Every time system
+reboot causes execution of **platform_reboot** script. This is platform specific
+script that prepares system for the reboot. One of the responsibility of that
+scripb is to make sure that ASIC firmware is up to date. To perform that check
+it calls **mlnx-fw-upgrade** script (for Nvidia platform) which actually
+performs checking of the versions (available and installed). In case versions
+are different it installs the ASIC firmware embed into system image
+(form */etc/mlnx* folder for Nvidia platform). See [Figure 1. ASIC firmware
+update original flow](#figure-1-asic-firmware-update-original-flow).
+
+<!-- omit in toc -->
+###### Figure 1. ASIC firmware update original flow
+<p align=center>
+  <img src="img/asic-firmware-update-orig-sequence.png"
+       alt="Figure 1. ASIC firmware update original flow">
+</p>
+
+When something went wrong during the reboot SONiC has a backup way. During the
+boottime **start.sh** script of **syncd** container calls the same
+**mlnx-fw-upgrade** script in order to make sure that ASIC firmware is up to
+date.
+
+#### New ASIC FW update flow
+The new ASIC firmware update flow brings couple of improvements to ASIC fw
+update script. It will read the configuration form the database and based on it
+decide if ASIC FW have to be updated.
+In the same manner the script will know the source of the fw image: system image
+or user defined.
+Example: for Nvidia platform system image ASIC firmware image located in
+*/etc/mlnx* directory. We bring a new user location for ASIC firmware image
+*/host/mlnx/asic*. This location will survive system upgrade, so system will be
+able to update ASIC FW even after system upgrade. See the new flow here:
+[Figure 2. ASIC firmware update controlled flow
+](#figure-2-asic-firmware-update-controlled-flow).
+
+<!-- omit in toc -->
+###### Figure 2. ASIC firmware update controlled flow
+<p align=center>
+  <img src="img/asic-firmware-update-new-sequence.png"
+       alt="Figure 2. ASIC firmware update controlled flow">
+</p>
+
+### Configuration and management
+#### CLI Enhancements
+This feature does not bring up any new CLI commands. But based on the caurrent
+SONiC CLI tree we can propose the next command tree.
+
+<!-- omit in toc -->
+#### CLI
+Proposed ASIC firmware control CLI tree:
+```
+config
```

Comment:
Did we explore using config platform firmware command to add ASIC?
config platform firmware
Usage: config platform firmware [OPTIONS] COMMAND [ARGS]...

  Firmware configuration tasks

Options:
  -h, -?, --help  Show this message and exit.

Commands:
  install  Install platform firmware
  update   Update platform firmware

Reply:
We can use that combination as well, but why don't use `config firmware asic`?
It will allow us to extend it with other components like `config firmware cpld` etc.

Reply:
config platform firmware already exists today. Adding asic will merely extend it. I don't think we can migrate "config platform firmware" commands to "config firmware" as it would break backward compatibility.

Reply:
Make sense. Updated